src/Algebra/Graph/Internal.hs

```diff
@@ -116,11 +116,12 @@ data Hit = Miss | Tail | Edge deriving (Eq, Ord)
 
 -- | A safe version of 'foldr1'.
 foldr1Safe :: (a -> a -> a) -> [a] -> Maybe a
-foldr1Safe f = foldr mf Nothing
-  where
-    mf x m = Just (case m of
-                        Nothing -> x
-                        Just y  -> f x y)
+foldr1Safe f = foldr (mf f) Nothing
+
+mf :: (a -> a -> a) ->  a -> Maybe a -> Maybe a
```

Comment:
Let's make the type more general `(a -> b -> a) -> a -> Maybe b -> Maybe a`, as it makes it a bit easier to understand what the function does. Also, I'd prefer to have a more meaningful name, and a comment :)

Reply:
Interesting: this type signature looks quite similar to `scanl`:

https://hackage.haskell.org/package/base-4.12.0.0/docs/Prelude.html#v:scanl

Reply:
Ok, this done, with changing its name to `maybeF`, and a better definition :)